guides/source/sign_up_and_settings.md

```diff
@@ -0,0 +1,1655 @@
+  def email_confirmation
+    @token = params[:user].generate_token_for(:email_confirmation)
+    mail to: params[:user].unconfirmed_email
+  end
+end
+```
+
+We'll include the token in the HTML view at `app/views/user_mailer/email_confirmation.html.erb`:
+
+```erb
+<h1>Verify your email address</h1>
+
+<p><%= link_to "Confirm your email", email_confirmation_url(token: @token) %></p>
+```
+
+And `app/views/user_mailer/email_confirmation.text.erb`:
+
+```erb
+Confirm your email: <%= email_confirmation_url(token: @token) %>
+```
+
+### Email Confirmation Controller
+
+The confirmation email includes a link to our Rails app to verify the email change.
+
+Let's add a route for this to `config/routes.rb`
+
+```ruby
+namespace :email do
+  resources :confirmations, param: :token, only: [ :show ]
+end
+```
+
+When a user clicks a link in their email, it will open a browser and make a GET request to the app. This means we only need the `show` action for this controller.
+
+Next, add the following to `app/controllers/emails/confirmations_controller.rb`
+
+```ruby
+class Email::ConfirmationsController < ApplicationController
+  allow_unauthenticated_access
+
+  def show
+    user = User.find_by_token_for(:email_confirmation, params[:token])
+    if user&.confirm_email
+      flash[:notice] = "Your email has been confirmed."
+    else
+      flash[:alert] = "Invalid token."
+    end
+    redirect_to root_path
+  end
+end
+```
+
+We want to confirm the email address whether the user is authenticated or not, so this controller allows unauthenticated access. Then we use the `find_by_token_for` method to validate the token and look up the matching `User` record. If successful, we update the user's email and reset `unconfirmed_email` to `nil` with the `confirm_email` method we added earlier. If the token isn't valid, the `user` variable will be `nil` so we can display an alert message in that case.
+
+Separating Admins & Users
+-------------------------
+
+Now that anyone can sign up for an account on our store, we need to differentiate between regular users and admins.
+
+### Adding An Admin Flag
+
+We'll start by adding a column to the User model.
+
+```bash
+$ rails g migration AddAdminToUsers admin:boolean
+```
+
+Then migrate the database.
+
+```bash
+$ rails db:migrate
+```
+
+A `User` with `admin` set to `true` should be able to add and remove products and access other administrative areas of the store.
+
+### Readonly Attributes
+
+We need to be very careful that `admin` is not editable by any malicious users. This is easy enough by keeping the `:admin` attribute out of any permitted parameters list.
+
+Optionally,we can mark the admin attribute as readonly for added security. This will tell Rails to raise an error anytime the admin attribute is changed. It can still be set when creating a record, but provides an additional layer of security. You may want to skip this if you'll be changing the admin flag for users often but in our e-commerce store it
```

Comment:
```suggestion
Optionally, we can mark the admin attribute as readonly for added security. This will tell Rails to raise an error anytime the admin attribute is changed. It can still be set when creating a record, but provides an additional layer of security against unauthorized changes. You may want to skip this if you'll be changing the admin flag for users often but in our e-commerce store, it's a useful safeguard.
```